test/apps/iodemo/io_demo.cc

```diff
@@ -360,64 +385,123 @@ class DemoClient : public P2pDemoCommon {
+        op_info_t curr[] = { { 0, 0 }, { 0, 0 } };
+        long total_iter  = 0;
 
-        long iteration = 0;
-        while ((iteration < opts().iter_count) && !_error_flag) {
-            VERBOSE_LOG << " <<<< iteration " << iteration << " >>>>";
+        while ((total_iter < opts().iter_count) && !_error_flag) {
+            VERBOSE_LOG << " <<<< iteration " << total_iter << " >>>>";
 
+            double prev_time = get_time();
             if (!wait_for_responses(opts().window_size - 1)) {
                 break;
             }
 
-            switch (opts().operation) {
+            io_op_t op = get_op();
+            size_t size;
+            switch (op) {
             case IO_READ:
-                do_io_read(conn, iteration);
+                size = do_io_read(conn, total_iter);
                 break;
             case IO_WRITE:
-                do_io_write(conn, iteration);
+                size = do_io_write(conn, total_iter);
                 break;
             default:
-                break;
+                abort();
             }
 
-            if (((iteration % 10) == 0) && (iteration > prev_iter)) {
-                double curr_time = get_time();
-                if (curr_time >= (prev_time + 1.0)) {
-                    if (!wait_for_responses(0)) {
-                        break;
-                    }
-
-                    report_performance(iteration - prev_iter,
-                                       curr_time - prev_time);
-                    prev_iter = iteration;
-                    prev_time = curr_time;
+            curr[op].data_size += size;
+            curr[op].time      += get_time() - prev_time;
+            curr[op].iter++;
+
+            if (((curr[op].iter % 10) == 0) && (curr[op].time >= 1.0)) {
+                if (!wait_for_responses(0)) {
+                    break;
                 }
+
+                size_t avg_data_size = curr[op].data_size / curr[op].iter;
+                report_performance(op, avg_data_size, curr[op].iter,
+                                   curr[op].time);
+
+                // start the new round
+                curr[op].data_size = 0;
+                curr[op].time      = 0;
+                curr[op].iter      = 0;
             }
 
-            ++iteration;
+            ++total_iter;
         }
 
         if (wait_for_responses(0)) {
-            report_performance(opts().iter_count - prev_iter,
-                               get_time() - prev_time);
+            size_t avg_data_size;
+
+            avg_data_size = curr[IO_READ].data_size / curr[IO_READ].iter;
+            report_performance(IO_READ, avg_data_size,
+                               curr[IO_READ].iter,
+                               curr[IO_READ].time);
+
+            avg_data_size = curr[IO_WRITE].data_size / curr[IO_WRITE].iter;
+            report_performance(IO_WRITE, avg_data_size,
+                               curr[IO_WRITE].iter,
+                               curr[IO_WRITE].time);
```

Comment:
can we unite these reports somehow to highlight that these ops potentially are done in parallel or pipelined

Reply:
done